### ISSUE TITLE
Sparse Tensor: in-place operation on detached tensors no longer raised error

### ISSUE THREAD

**Issue by HanGuo97** [pytorch/pytorch#98976]
### 🐛 Describe the bug

According to the [documentation](https://pytorch.org/docs/stable/generated/torch.Tensor.detach.html), in-place operations to detached sparse tensors would raise an error. However, it did not. This has implications for modifying the `.state_dict` (which detaches the tensor by default) in place.

### Versions

[pip3] mypy==1.1.1
[pip3] mypy-extensions==1.0.0
[pip3] numpy==1.23.5
[pip3] torch==2.0.0
[pip3] torchaudio==2.0.0
[pip3] torchdata==0.6.0
[pip3] torchelastic==0.2.2
[pip3] torchtext==0.15.0
[pip3] torchvision==0.15.0
[pip3] triton==2.0.0
[conda] blas                      1.0                         mkl  
[conda] ffmpeg                    4.3                  hf484d3e_0    pytorch
[conda] magma-cuda117             2.6.1                         1    pytorch
[conda] mkl                       2021.4.0           h06a4308_640  
[conda] mkl-service               2.4.0           py310h7f8727e_0  
[conda] mkl_fft                   1.3.1           py310hd6ae3a3_0  
[conda] mkl_random                1.2.2           py310h00e6091_0  
[conda] numpy                     1.23.5          py310hd5efca6_0  
[conda] numpy-base                1.23.5          py310h8e6c178_0  
[conda] pytorch                   2.0.0           py3.10_cuda11.7_cudnn8.5.0_0    pytorch
[conda] pytorch-cuda              11.7                 h778d358_3    pytorch
[conda] pytorch-mutex             1.0                        cuda    pytorch
[conda] torchaudio                2.0.0               py310_cu117    pytorch
[conda] torchdata                 0.6.0                     py310    pytorch
[conda] torchelastic              0.2.2                    pypi_0    pypi
[conda] torchtext                 0.15.0                    py310    pytorch
[conda] torchtriton               2.0.0                     py310    pytorch
[conda] torchvision               0.15.0              py310_cu117    pytorch

cc @alexsamardzic @nikitaved @pearu @cpuhrsch @amjames @bhosmer

**Comment by dagitses** [comment]
Can you give a repro please?

**Comment by HanGuo97** [comment]
Sure.

```python
import torch
A = torch.eye(9).to_sparse_coo()
B = torch.zeros_like(A)
A.detach().copy_(B)
print(A)
```

**Comment by cpuhrsch** [comment]
@pearu - Can you take a look?

**Comment by albanD** [comment]
Ho yes that docs may very well need to be updated to reflect the latest behavior: metadata change don't raise errors anymore.
For sparse, I'm not sure how `.values().zero_()` could ever have triggered an error tbh. But if that's not accurate we should update the doc yes.
